packages/cozy-client/src/manifest.js

```diff
@@ -84,5 +114,112 @@ export function sanitize(manifest) {
+ *
+ * @param  {Object} [fields={}] Manifest fields
+ * @returns {Object}             Sanitized manifest fields
+ */
+export const sanitizeIdentifier = fields => {
+  const sanitized = _cloneDeep(fields)
+  let hasIdentifier = false
+  for (let fieldName in sanitized)
+    if (sanitized[fieldName].role === ROLE_IDENTIFIER) {
+      if (hasIdentifier) delete sanitized[fieldName].role
+      else hasIdentifier = true
+    }
+  if (hasIdentifier) return sanitized
+
+  for (let name of legacyLoginFields)
+    if (sanitized[name]) {
+      sanitized[name].role = ROLE_IDENTIFIER
+      return sanitized
+    }
+
+  for (let fieldName in sanitized)
+    if (sanitized[fieldName].type !== 'password') {
+      sanitized[fieldName].role = ROLE_IDENTIFIER
+      return sanitized
+    }
+
+  return sanitized
+}
+
+/**
+ * Returns the key for the field having the role=identifier attribute
+ *
+ * @param  {Object} fields Konnector fields
+ * @returns {string}        The key for the identifier field, example 'login'
+ */
+export const getIdentifier = (fields = {}) =>
+  findKey(sanitizeIdentifier(fields), field => field.role === ROLE_IDENTIFIER)
+/**
+ * Ensures old fields are removed
+ *
+ * @param  {Object} fields Manifest fields
+ * @returns {Object}        Sanitized manifest fields
+ */
+const removeOldFields = fields => {
+  const sanitized = _cloneDeep(fields)
+  delete sanitized.advancedFields
+  return sanitized
+}
+
+/**
+ * Ensures every field not explicitely tagged as not required is required
+ *
+ * @param  {Object} [fields={}] Manifest fields
+ * @returns {Object}             Sanitized manifest fields
+ */
+const sanitizeRequired = fields => {
+  const sanitized = _cloneDeep(fields)
+  for (let fieldName in sanitized) {
+    const field = sanitized[fieldName]
+    // Ensure legacy for field isRequired
+    const required =
+      typeof field.required === 'undefined' ? field.isRequired : field.required
+    sanitized[fieldName].required =
+      typeof required === 'boolean' ? required : true
+  }
+
+  return sanitized
+}
+
+/**
+ * Ensures:
+ * * any field flagged as encrypted keeps its flag
+ * * any legacy encrypted field is tagged as encrypted
+ *
+ * @param  {Object} [fields={}] Manifest fields
+ * @returns {Object}             Sanitized Manifest fields
+ */
+const sanitizeEncrypted = fields => {
+  const sanitized = _cloneDeep(fields)
+  for (let fieldName in sanitized) {
+    const field = sanitized[fieldName]
+    if (typeof field.encrypted !== 'boolean')
+      field.encrypted =
+        field.type === 'password' || legacyEncryptedFields.includes(fieldName)
+  }
+  return sanitized
+}
+
+/**
+ * Sanitizes manifest fields with multiple rules
+ *
+ * @param  {Object} [fields={}] Manifest fields
+ * @returns {Object}            Sanitized manifest fields
+ */
+const sanitizeFields = _flow([
+  removeOldFields,
+  sanitizeIdentifier,
+  sanitizeRequired,
+  sanitizeEncrypted
+])
```

Comment:
I didn't even know about this file. I think we should move it to /models directory. Can you move it and create a BC for that?

Reply:
Done, but this actually not a BC since the manifest lib is called via client.manifest and I fixed index.js and index.node.js. Don't you think ?

Reply:
> Done, but this actually not a BC since the manifest lib is called via client.manifest and I fixed index.js and index.node.js. Don't you think ?

We've a lot of deep import since our tree shaking is not working correctly, so yes I consider it as a BC.